File: Graph/java/LargestComponent.java
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class LargestComponent {
  public static int size(Map<String, List<String>> graph) {
    Set<String> visited = new HashSet<>();
    int largest = 0;

    for (String node : graph.keySet()) {
      int current = exploreSize(graph, node, visited);
      if (current > largest) {
        largest = current;
      }
    }

    return largest;
  }

  private static int exploreSize(Map<String, List<String>> graph, String node, Set<String> visited) {
    if (!visited.add(node)) {
      return 0;
    }

    int size = 1;
    for (String neighbor : graph.getOrDefault(node, List.of())) {
      size += exploreSize(graph, neighbor, visited);
    }
    return size;
  }
}

File: Graph/csharp/LargestComponent.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class LargestComponent
{
    public static int Solve(Dictionary<string, List<string>> graph)
    {
        var visited = new HashSet<string>();
        var largest = 0;

        foreach (var node in graph.Keys)
        {
            var size = ExploreSize(graph, node, visited);
            if (size > largest)
            {
                largest = size;
            }
        }

        return largest;
    }

    private static int ExploreSize(Dictionary<string, List<string>> graph, string node, HashSet<string> visited)
    {
        if (!visited.Add(node))
        {
            return 0;
        }

        if (!graph.TryGetValue(node, out var neighbors))
        {
            return 1;
        }

        var size = 1;
        foreach (var neighbor in neighbors)
        {
            size += ExploreSize(graph, neighbor, visited);
        }

        return size;
    }
}
